### ISSUE TITLE
Earthquake 

### ISSUE THREAD

**Issue by BaNru** [CorsixTH/CorsixTH#1462]
#### Describe the issue
Earthquake. See the operating theater. When an explosion occurs, an error occurs.

#### Steps to Reproduce
1. Load save game attached to this ticket
2. Wait for the earthquake
3. We look at the error

#### Save Game
https://www.dropbox.com/s/cd7ag8mh42ldu48/_Autosave3.zip

#### System Information
**CorsixTH Version: 0.62 from https://launchpad.net/ubuntu/+source/corsix-th/0.62-1/+build/15258578
**Operating System: kUbuntu
**Theme Hospital Version: different sets

#### Gamelog.txt
![Gamelog](http://dl3.joxi.net/drive/2018/09/10/0004/2123/305227/27/dd76471233.png)

### Additional Info
![Screen](http://dl3.joxi.net/drive/2018/09/10/0004/2123/305227/27/fb039ded9d.png)

**Comment by mugmuggy** [comment]
There appears to be a few annoying gremlins with timing here. This error is thrown when the swing door is ticking at the time of the room crash. Easy fix is to put a guard in Room:tryAdvanceQueue for example to check that queue is not nil. Like so
```lua
function Room:tryAdvanceQueue()
  if self.door.queue and self.door.queue:size() > 0 and not self.door.user and not self.door.reserved_for then
```
But I'm unable to reliably hit this one issue. If the door is in use a bit more, the swing animation is still in progress after the objects in the room are destroyed. So maybe also in closeDoor if its ticking, set timer so door animations can switch back on next tick.

Sometimes I get another one where the patient tries to enter the room after the crash. As a work around I just put in Door:closeDoor() which will interrupt the walk. Hunch is this is related to the drinks machine/must_happen issue.
```lua
  local map = self.world.map
  local x, y = self.room:getEntranceXY(true)
  map:setCellFlags(x, y, {passable = false})
```
But I also see some weird walk behaviour from patients - for example the bench near the drinks machine. Patients will enter from the west side and 'into' the bench tile. But after the crash room, they may also go out the same way, then 'teleport' to in front of doors or benches further up.

**Comment by metsatron** [comment]
![screenshot_2018-11-01_01-00-54](https://user-images.githubusercontent.com/3039914/47825375-d20e5e00-dd71-11e8-8621-0dd556721323.png)

I believe I'm having the same if not a similar problem. Is there any way for now to temporarily disable earthquakes?

**Comment by mugmuggy** [comment]
That is probably #1487, the easiest workaround is to pick them up and place the handyman in the room, but that room is exploding so you probably want to replace the machine to prevent the room crash.

**Comment by mugmuggy** [comment]
Sounds like the door reservation bug, and they might be coming in to water plants. It is how I replicated #1487. In that instance editing the room and destroying the door, destroys the reservation. Feel free to upload save to #1487 and I can take a look.

**Comment by jonathan-gramain** [comment]
FWIW, I ran into a somewhat similar stacktrace, that led to the operation room being stuck. I compiled from source with commit c97629ed.

```
A stack trace is included below, and the handler has been disconnected.
...spital/CorsixTH/CorsixTH/Lua/rooms/operating_theatre.lua:181: attempt to compare number with nil
stack traceback:
        ...spital/CorsixTH/CorsixTH/Lua/rooms/operating_theatre.lua:181: in method 'getStaffMember'
        ...jeux/dosbox/ThemeHospital/CorsixTH/CorsixTH/Lua/room.lua:333: in method 'onHumanoidEnter'
        ...spital/CorsixTH/CorsixTH/Lua/humanoid_actions/pickup.lua:51: in local 'on_interrupt'
        ...hemeHospital/CorsixTH/CorsixTH/Lua/entities/humanoid.lua:594: in method 'setNextAction'
        ...meHospital/CorsixTH/CorsixTH/Lua/dialogs/place_staff.lua:60: in method 'close'
        ...meHospital/CorsixTH/CorsixTH/Lua/dialogs/place_staff.lua:140: in method 'onMouseUp'
        ...ux/dosbox/ThemeHospital/CorsixTH/CorsixTH/Lua/window.lua:1462: in field 'onMouseUp'
        ...n/jeux/dosbox/ThemeHospital/CorsixTH/CorsixTH/Lua/ui.lua:673: in function <...n/jeux/dosbox/ThemeHospital/CorsixTH/CorsixTH/Lua/ui.lua:660>
        (...tail calls...)
        .../jeux/dosbox/ThemeHospital/CorsixTH/CorsixTH/Lua/app.lua:915: in function <.../jeux/dosbox/ThemeHospital/CorsixTH/CorsixTH/Lua/app.lua:910>
```

**Comment by mugmuggy** [comment]
Just expanding on earlier comment, https://github.com/CorsixTH/CorsixTH/issues/1462#issuecomment-420626354 the last bit regarding the 'teleporting'.

Drinks machine appears to do this when queued patient is walking to the drinks machine and the room crashes (or probably anything else that interrupts the walk), the UseObjectAction then starts the animation from the patients current position and not the drinks machine.

The QueueAction, action_queue_get_soda allows truncatation of the walk, that leads to this, where as Patient:tickDay disables the truncate.

The same process occurs with the bench I assume except the patient 'jumps' instead to the bench when UseObjectAction runs rather than the bench 'jumping' to them.

**Comment by mugmuggy** [comment]
Original issue fixed with #1566.
